library/ecp.c

```diff
@@ -3125,10 +3682,76 @@ int mbedtls_ecp_gen_privkey( const mbedtls_ecp_group *grp,
     if( mbedtls_ecp_get_type( grp ) == MBEDTLS_ECP_TYPE_SHORT_WEIERSTRASS )
         return( mbedtls_ecp_gen_privkey_sw( &grp->N, d, f_rng, p_rng ) );
 #endif /* MBEDTLS_ECP_SHORT_WEIERSTRASS_ENABLED */
+#if defined(MBEDTLS_ECP_EDWARDS_ENABLED)
+    if( mbedtls_ecp_get_type( grp ) == MBEDTLS_ECP_TYPE_EDWARDS )
+        return( mbedtls_ecp_gen_privkey_mx( grp->nbits, d, f_rng, p_rng ) );
+#endif /* MBEDTLS_ECP_EDWARDS_ENABLED */
 
     return( MBEDTLS_ERR_ECP_BAD_INPUT_DATA );
 }
 
+#if defined(MBEDTLS_ECP_EDWARDS_ENABLED)
+#if defined(MBEDTLS_ECP_DP_ED25519_ENABLED)
+static int mbedtls_ecp_expand_ed25519( const mbedtls_mpi *d,
+                    mbedtls_mpi *q, mbedtls_mpi *prefix )
+{
+    int ret = 0;
+
+    if( mbedtls_mpi_size( d ) != 32)
+        return( MBEDTLS_ERR_ECP_INVALID_KEY );
+
+    unsigned char key_buf[32], sha_buf[64];
+    MBEDTLS_MPI_CHK( mbedtls_mpi_write_binary_le( d, key_buf, sizeof( key_buf ) ) );
+
+    mbedtls_sha512( key_buf, sizeof( key_buf ), sha_buf, 0 );
+
+    sha_buf[0] &= ~0x7;
+    sha_buf[31] &= ~0x80;
+    sha_buf[31] |= 0x40;
+
+    MBEDTLS_MPI_CHK( mbedtls_mpi_read_binary_le( q, sha_buf, 32 ) );
+    if( prefix )
+    {
+        MBEDTLS_MPI_CHK( mbedtls_mpi_read_binary_le( prefix, sha_buf + 32, 32 ) );
+    }
+
+cleanup:
```

Comment:
`key_buf` contains the private key and `sha_buf` the corresponding ephemeral key. They should not be left on the stack, you must use `mbedtls_platform_zeroize` to ensure they are cleared.